README
SNMP-LogParser

The SNMP-LogParser is one component script that allows to access log
file information via SNMP. 

The logparser script is used to incrementally parse log files.
The idea is that the log parser is invoked from Unix cron updating a stats file that
could then be tied easily into Net-SNMP (see http://www.net-snmp.org) by using the
SNMP-MibProxy module.

This module has bit of learning curve, you need to learn the methods of the base
class B<SNMP::LogParserDriver> and then implement an inherited class of
it to the parsing that you need.

Once you've learned it, you can do quick deployments of SNMP info gathered from
log files.

You have more examples of LogParser Drivers in the EXAMPLES directory

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

Please be aware that the installation tries to write create the directories:
/etc/logparser, /var/log/logparser and /var/lib/logparser.
To change those directories please run:

    perl Build.PL --install_path etc=/etc/logparser --install_path log=/var/log/logparser --install_path work=/var/lib/logparser 


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc SNMP::LogParser

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/SNMP-LogParser

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=SNMP-LogParser

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/SNMP-LogParser

    CPAN Ratings:
        http://cpanratings.perl.org/d/SNMP-LogParser

COPYRIGHT AND LICENCE

Copyright (C) 2007 Nito Martinez

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
$Id: Changes 15291 2012-05-03 21:01:14Z rporres $

Revision history for SNMP-LogParser

1.15290 03/05/2012

   [BUGS]

   * Fixed bug in include option

1.14996  10/04/2012

    [OTHER]

    * Copied lock.t test to concurrentProcess.t which was empty, waiting to be done
      and has a more descriptive name than lock.t

1.14990  10/04/2012

    [ENHANCEMENTS]

    * Added lock test

    [OTHER]

    * Cleaned-up code to follow modern practices
    * Will work only on perl version >= 5.6 as needed by required modules
    * Reorganized examples

1.10178  10/02/2011
         
    [ENHANCEMENTS]

    * Added -I command line option to add directories to @INC (Rafael Porres - rporres)

    [BUG FIXES]

    * Command line options -k and -o didn't work as advertised (rporres)

1.10052  25/1/2011
	 Updated eval condition on non existent logdrivers

1.0767  1/12/2007
        Changed name from NetSNMP-LogParser to SNMP-LogParser. Suggestion from brian d foy <brian.d.foy@gmail.com>  (Perl Authors Upload Server)

1.0537  18/10/2007
        Changed SEEK_CUR in &Fcntl::SEEK_CUR because of "t/logparser............Undefined subroutine &File::Temp::SEEK_SET called at blib/script/logparser line 332."


1.0530  17/10/2007
        Removed taint mode in t/logparser.t (Problems with File::Path 2.01 in perl 5.6.2). I rather have it working in more versions than tainted

1.0481  15/10/2007
        Changed Build.PL for a dependency on File::Path 1.05 for Taint mode

1.0480  14/10/2007
        Changed t/mibProxy.t to not setup PERL5LIB

1.0479  14/10/2007
        Changed lib/NetSNMP/MibProxy.pm to not use the our declaration which breaks the test on perl5.005

1.0475    14-10-2007
	Delete the only used once $opt_d in line 162 of bin/logparser
	Added Test::Pod::Coverage 1.04 to the requisites in Build.PL

1.0469    11-10-2007
        Changed name from QND-NetSNMP-LogParser to NetSNMP-LogParser


1.0462    5-10-2007
        First public version



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3607
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'SNMP::LogParser',
  'VERSION_FROM' => 'lib/SNMP/LogParser.pm',
  'PREREQ_PM' => {
                   'Config::Properties::Simple' => '0.14',
                   'File::Path' => '1.05',
                   'File::Spec' => 0,
                   'File::Temp' => 0,
                   'File::stat' => 0,
                   'IPC::Run3' => 0,
                   'Log::Dispatch::FileRotate' => 0,
                   'Log::Log4perl' => 0,
                   'POSIX' => 0,
                   'Pod::Usage' => 0,
                   'Proc::ProcessTable' => 0,
                   'Statistics::Descriptive' => 0,
                   'Storable' => 0,
                   'Test::More' => 0,
                   'Test::Pod' => '1.14',
                   'Test::Pod::Coverage' => '1.04',
                   'Test::SharedFork' => 0,
                   'lib' => 0,
                   'parent' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [
                   'bin/logparser'
                 ],
  'PL_FILES' => {}
)
;


EXAMPLES/ProxyLog.pm
#
# $Id: ProxyLog.pm 14990 2012-04-10 12:57:24Z rporres $
#
# A bit more complicated class
#
# Nito@Qindel.ES -- 7/9/2006
package SNMP::LogParserDriver::ProxyLog;
use strict;
use warnings;
use parent 'SNMP::LogParserDriver';

our %nsCacheIPClient;
our %nsCacheUserClient;
our %nsCacheURL;
our @nsCacheIPClient;
our @nsCacheUserClient;
our @nsCacheURL;

# Class constructor
sub new {
  my $class = shift;
  my $self  = $class->SUPER::new();
  bless ($self, $class);
  $self->pattern('^(\d+\.\d+\.\d+\.\d+)\s\S+\s(\S+)\s\[[^]]+\]\s\"[^"]+\"\s(\S+)\s(\S+)\s\"[^"]+\"\s\"[^"]+\"\s\S+\s(\S+)\s\S+\s\"[^"]+\"\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s\S+\s\S+\s\S+\s\S+\s\S+\s\S+\s(\S+)\s\S+$');
  return $self;
}


# This will be invoked before the first parsing of the log
sub evalBegin {
  my $self = shift;
  for my $name ('nsCacheClientRequests', 'nsCacheClientSuccess',
		'nsCacheClientErrors', 'nsCacheClientNoStatus',
                'nsCacheClientInByte', 'nsCacheClientOutByte',
		'nsCacheServerRequests', 'nsCacheServerSuccess',
                'nsCacheServerErrors', 'nsCacheServerNoStatus',
		'nsCacheServerInByte', 'nsCacheServerOutByte',
		'nsCacheClientResponseTimeEntries',
		'nsCacheClientAvgResponseTime',
		'nsCacheClientStdDeviationResponseTime',
		'nsCacheClientMinResponseTime',
		'nsCacheClientMaxResponseTime',
		'nsCacheClientDNSResponseTimeEntries',
		'nsCacheClientAvgDNSResponseTime',
		'nsCacheClientStdDeviationDNSResponseTime',
		'nsCacheClientMinDNSResponseTime',
		'nsCacheClientMaxDNSResponseTime' )
  {
    $self->{savespace}{$name} = 0 if (!defined($self->{savespace}{$name}));
  }
  $self->{savespace}{'responseTime'} = Statistics::Descriptive::Sparse->new();
  $self->{savespace}{'dnsResponseTime'} = Statistics::Descriptive::Sparse->new();
}

### This will be invoked whenever the pattern matches
### the log line parsed
### Input:
### - The line to be parsed
### Output:
### - 1 if the line has matched the regular expression and 0 otherwise
sub evalIterate {
  my $self = shift;
  my ($line) = @_;
  my $pattern = $self->{pattern};
  if ($line =~ /$pattern/) {
    my ($ip, $user, $url, $serverStatus, $remServerStatus) = ($1, $2, $5, $3, $6);
    my ($CLp2c, $CLr2p, $CLc2p, $CLp2r, $HLc2p, $HLp2c, $HLp2r, $HLr2p) =
      ($4, $7, $8, $9, $10, $11, $12, $13);
    my ($totalTime, $dnsTime) = ($14, $15) ;
    if ($url =~ /^([^:]+:\/\/[^\/]+)\//) {
      $url = $1;
    }
    if (!exists($nsCacheIPClient{$ip})) {
      $nsCacheIPClient{$ip}{'Addr'} = $ip ;
      $nsCacheIPClient{$ip}{'Requests'} = 0 ;
      $nsCacheIPClient{$ip}{'Success'} = 0 ;
      $nsCacheIPClient{$ip}{'Error'} = 0 ;
      $nsCacheIPClient{$ip}{'NoStatus'} = 0 ;
      $nsCacheIPClient{$ip}{'InByte'} = 0 ;
      $nsCacheIPClient{$ip}{'OutByte'} = 0 ;
      $nsCacheIPClient{$ip}{'ServerInByte'} = 0 ;
      $nsCacheIPClient{$ip}{'ServerOutByte'} = 0 ;
      $nsCacheIPClient{$ip}{'ResponseTime'} = Statistics::Descriptive::Sparse->new();
    }
    if (!exists($nsCacheUserClient{$user})) {
      $nsCacheUserClient{$user}{'Uid'} = $user ;
      $nsCacheUserClient{$user}{'Requests'} = 0 ;
      $nsCacheUserClient{$user}{'Success'} = 0 ;
      $nsCacheUserClient{$user}{'Error'} = 0 ;
      $nsCacheUserClient{$user}{'NoStatus'} = 0 ;
      $nsCacheUserClient{$user}{'InByte'} = 0 ;
      $nsCacheUserClient{$user}{'OutByte'} = 0 ;
      $nsCacheUserClient{$user}{'ServerInByte'} = 0 ;
      $nsCacheUserClient{$user}{'ServerOutByte'} = 0 ;
      $nsCacheUserClient{$user}{'ResponseTime'} = Statistics::Descriptive::Sparse->new();
    }
    if (!exists($nsCacheURL{$url})) {
      $nsCacheURL{$url}{'String'} = $url ;
      $nsCacheURL{$url}{'Requests'} = 0 ;
      $nsCacheURL{$url}{'Success'} = 0 ;
      $nsCacheURL{$url}{'Error'} = 0 ;
      $nsCacheURL{$url}{'NoStatus'} = 0 ;
      $nsCacheURL{$url}{'InByte'} = 0 ;
      $nsCacheURL{$url}{'OutByte'} = 0 ;
      $nsCacheURL{$url}{'ServerInByte'} = 0 ;
      $nsCacheURL{$url}{'ServerOutByte'} = 0 ;
      $nsCacheURL{$url}{'ResponseTime'} = Statistics::Descriptive::Sparse->new();
    }
    $nsCacheIPClient{$ip}{'Requests'} ++;
    $nsCacheUserClient{$user}{'Requests'} ++;
    $nsCacheURL{$url}{'Requests'} ++;
    $self->{savespace}{'nsCacheClientRequests'} ++;
    if ($serverStatus =~ /\d+/) {
      if ($serverStatus <= 399) {
	$self->{savespace}{'nsCacheClientSuccess'} ++;
	$nsCacheIPClient{$ip}{'Success'} ++;
	$nsCacheUserClient{$user}{'Success'} ++;
	$nsCacheURL{$url}{'Success'} ++;
      }
      else {
	$self->{savespace}{'nsCacheClientErrors'} ++;
	$nsCacheIPClient{$ip}{'Error'} ++;
	$nsCacheUserClient{$user}{'Error'} ++;
	$nsCacheURL{$url}{'Error'} ++;
      }
    }
    else {
      $self->{savespace}{'nsCacheClientNoStatus'} ++;
      $nsCacheIPClient{$ip}{'NoStatus'} ++;
      $nsCacheUserClient{$user}{'NoStatus'} ++;
      $nsCacheURL{$url}{'NoStatus'} ++;
    }
    if ($remServerStatus =~ /\d+/)
      { if ($remServerStatus <= 399 )
	  {  $self->{savespace}{'nsCacheServerSuccess'} ++; }
	else
	  { $self->{savespace}{'nsCacheServerErrors'} ++; }
      }
    else
      { $self->{savespace}{'nsCacheServerNoStatus'} ++; }
    $CLc2p = 0 if ($CLc2p eq '-');
    $HLc2p = 0 if ($HLc2p eq '-');
    $CLp2c = 0 if ($CLp2c eq '-');
    $HLp2c = 0 if ($HLp2c eq '-');
    $CLr2p = 0 if ($CLr2p eq '-');
    $HLr2p = 0 if ($HLr2p eq '-');
    $CLp2r = 0 if ($CLp2r eq '-');
    $HLp2r = 0 if ($HLp2r eq '-');
    $self->{savespace}{'nsCacheClientInByte'} += $CLc2p + $HLc2p;
    $nsCacheIPClient{$ip}{'InByte'} += $CLc2p + $HLc2p;
    $nsCacheUserClient{$user}{'InByte'} += $CLc2p + $HLc2p;
    $nsCacheURL{$url}{'InByte'} += $CLc2p + $HLc2p;
    $self->{savespace}{'nsCacheClientOutByte'} += $CLp2c + $HLp2c;
    $nsCacheIPClient{$ip}{'OutByte'} += $CLp2c + $HLp2c;
    $nsCacheUserClient{$user}{'OutByte'} += $CLp2c + $HLp2c;
    $nsCacheURL{$url}{'OutByte'} += $CLp2c + $HLp2c;
    $self->{savespace}{'nsCacheServerInByte'} += $CLr2p + $HLr2p;
    $nsCacheIPClient{$ip}{'ServerInByte'} += $CLr2p + $HLr2p;
    $nsCacheUserClient{$user}{'ServerInByte'} += $CLr2p + $HLr2p;
    $nsCacheURL{$url}{'ServerInByte'} += $CLr2p + $HLr2p;
    $self->{savespace}{'nsCacheServerOutByte'} += $CLp2r + $HLp2r;
    $nsCacheIPClient{$ip}{'ServerOutByte'} += $CLp2r + $HLp2r;
    $nsCacheUserClient{$user}{'ServerOutByte'} += $CLp2r + $HLp2r;
    $nsCacheURL{$url}{'ServerOutByte'} += $CLp2r + $HLp2r;
    if ($totalTime =~ /\d+/) {
      $self->{savespace}{'responseTime'}->add_data($totalTime * 1000);
      $nsCacheIPClient{$ip}{'ResponseTime'}->add_data($totalTime * 1000);
      $nsCacheUserClient{$user}{'ResponseTime'}->add_data($totalTime * 1000);
      $nsCacheURL{$url}{'ResponseTime'}->add_data($totalTime * 1000);
    }
    if ($dnsTime =~ /\d+/)
      { $self->{savespace}{'dnsResponseTime'}->add_data($dnsTime * 1000); }
  }
}


### This will be invoked after the last log line has been parsed
sub evalEnd {
  my $self = shift;
  my $rt = $self->{savespace}{'responseTime'};
  my $dt = $self->{savespace}{'dnsResponseTime'};
  if ($rt->count() != 0)
    {
      $self->{savespace}{'nsCacheClientResponseTimeEntries'} = $rt->count();
      $self->{savespace}{'nsCacheClientAvgResponseTime'} = $rt->mean();
      $self->{savespace}{'nsCacheClientStdDeviationResponseTime'} = defined($rt->standard_deviation()) ? $rt->standard_deviation() : 0;
      $self->{savespace}{'nsCacheClientMinResponseTime'} = $rt->min();
      $self->{savespace}{'nsCacheClientMaxResponseTime'} = $rt->max();
    }
  if ($dt->count() != 0)
    {
      $self->{savespace}{'nsCacheClientDNSResponseTimeEntries'} = $dt->count();
      $self->{savespace}{'nsCacheClientAvgDNSResponseTime'} = $dt->mean();
      $self->{savespace}{'nsCacheClientStdDeviationDNSResponseTime'} = defined($dt->standard_deviation()) ? $dt->standard_deviation() : 0;
      $self->{savespace}{'nsCacheClientMinDNSResponseTime'} = $dt->min();
      $self->{savespace}{'nsCacheClientMaxDNSResponseTime'} = $dt->max();
    }
  my $maxArrayIndex = 50;
  $self->{logger}->debug("Starting sort of IPClient");
  use Tie::Array::Sorted;
  tie @nsCacheIPClient, "Tie::Array::Sorted", sub { ${$_[1]}{'Requests'} <=> ${$_[0]}{'Requests'} };
  foreach my $k (keys %nsCacheIPClient)
    {
      if ($#nsCacheIPClient < $maxArrayIndex  - 1)
	{
	  push @nsCacheIPClient, $nsCacheIPClient{$k};
	}
      else
	{
	  if (${$nsCacheIPClient[$#nsCacheIPClient]}{'Requests'} < $nsCacheIPClient{$k}{'Requests'})
	    {
	      delete $nsCacheIPClient[$#nsCacheIPClient];
	      push @nsCacheIPClient, $nsCacheIPClient{$k};
	    }
	}
    }
  $self->{logger}->debug("End of sort of IPClient");
  $self->{savespace}{'nsCacheIPClientTableEntries'} = $#nsCacheIPClient + 1;
  for (my $i = 0 ; $i <= $#nsCacheIPClient; $i ++) {
    my $i1 = $i + 1;
    $self->{savespace}{'nsCacheIPClientIndex.'.$i1} = $i1;
    foreach my $k (keys %{$nsCacheIPClient[$i]}) {
      if ($k ne 'ResponseTime') {
	$self->{savespace}{'nsCacheIPClient' . $k . '.'. $i1} = ${$nsCacheIPClient[$i]}{$k};
      } else {
	if (${$nsCacheIPClient[$i]}{$k}->count == 0) {
	  $self->{savespace}{'nsCacheIPClientResponseTimeEntries'. '.'. $i1} = 0;
	  $self->{savespace}{'nsCacheIPClientAvgResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheIPClientStdDeviationResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheIPClientMinResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheIPClientMaxResponseTime'. '.'. $i1} =  0;
	} else {
	  $self->{savespace}{'nsCacheIPClientResponseTimeEntries'. '.'. $i1} = ${$nsCacheIPClient[$i]}{'ResponseTime'}->count();
	  $self->{savespace}{'nsCacheIPClientAvgResponseTime'. '.'. $i1} = ${$nsCacheIPClient[$i]}{'ResponseTime'}->mean();
	  $self->{savespace}{'nsCacheIPClientStdDeviationResponseTime'. '.'. $i1} = defined(${$nsCacheIPClient[$i]}{'ResponseTime'}->standard_deviation()) ? ${$nsCacheIPClient[$i]}{'ResponseTime'}->standard_deviation() : 0;
	  $self->{savespace}{'nsCacheIPClientMinResponseTime'. '.'. $i1} = ${$nsCacheIPClient[$i]}{'ResponseTime'}->min();
	  $self->{savespace}{'nsCacheIPClientMaxResponseTime'. '.'. $i1} = ${$nsCacheIPClient[$i]}{'ResponseTime'}->max();
	}
      }
    }
  }
  $self->{logger}->debug("Starting sort of UserClient");
  use Tie::Array::Sorted;
  tie @nsCacheUserClient, "Tie::Array::Sorted", sub { ${$_[1]}{'Requests'} <=> ${$_[0]}{'Requests'} };
  foreach my $k (keys %nsCacheUserClient)
    {
      if ($#nsCacheUserClient < $maxArrayIndex - 1 )
	{
	  push @nsCacheUserClient, $nsCacheUserClient{$k};
	}
      else
	{
	  if (${$nsCacheUserClient[$#nsCacheUserClient]}{'Requests'} < $nsCacheUserClient{$k}{'Requests'})
	    {
	      delete $nsCacheUserClient[$#nsCacheUserClient];
	      push @nsCacheUserClient, $nsCacheUserClient{$k};
	    }
	}
    }
  $self->{logger}->debug("End of sort of UserClient");
  $self->{savespace}{'nsCacheUserClientTableEntries'} = $#nsCacheUserClient + 1;
  for (my $i = 0 ; $i <= $#nsCacheUserClient; $i ++) {
    my $i1 = $i + 1;
    $self->{savespace}{'nsCacheUserClientIndex.'.$i1} = $i1;
    foreach my $k (keys %{$nsCacheUserClient[$i]}) {
      if ($k ne 'ResponseTime') {
	$self->{savespace}{'nsCacheUserClient' . $k . '.'. $i1} = ${$nsCacheUserClient[$i]}{$k};
      } else {
	if (${$nsCacheUserClient[$i]}{$k}->count == 0) {
	  $self->{savespace}{'nsCacheUserClientResponseTimeEntries'. '.'. $i1} = 0;
	  $self->{savespace}{'nsCacheUserClientAvgResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheUserClientStdDeviationResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheUserClientMinResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheUserClientMaxResponseTime'. '.'. $i1} =  0;
	} else {
	  $self->{savespace}{'nsCacheUserClientResponseTimeEntries'. '.'. $i1} = ${$nsCacheUserClient[$i]}{'ResponseTime'}->count();
	  $self->{savespace}{'nsCacheUserClientAvgResponseTime'. '.'. $i1} = ${$nsCacheUserClient[$i]}{'ResponseTime'}->mean();
	  $self->{savespace}{'nsCacheUserClientStdDeviationResponseTime'. '.'. $i1} = defined(${$nsCacheUserClient[$i]}{'ResponseTime'}->standard_deviation()) ? ${$nsCacheUserClient[$i]}{'ResponseTime'}->standard_deviation() : 0;
	  $self->{savespace}{'nsCacheUserClientMinResponseTime'. '.'. $i1} = ${$nsCacheUserClient[$i]}{'ResponseTime'}->min();
	  $self->{savespace}{'nsCacheUserClientMaxResponseTime'. '.'. $i1} = ${$nsCacheUserClient[$i]}{'ResponseTime'}->max();
	}
      }
    }
  }
  $self->{logger}->debug("Starting sort of URL");
  use Tie::Array::Sorted;
  tie @nsCacheURL, "Tie::Array::Sorted", sub { ${$_[1]}{'Requests'} <=> ${$_[0]}{'Requests'} };
  foreach my $k (keys %nsCacheURL)
    {
      if ($#nsCacheURL < $maxArrayIndex - 1)
	{
	  push @nsCacheURL, $nsCacheURL{$k};
	}
      else
	{
	  if (${$nsCacheURL[$#nsCacheURL]}{'Requests'} < $nsCacheURL{$k}{'Requests'})
	    {
	      delete $nsCacheURL[$#nsCacheURL];
	      push @nsCacheURL, $nsCacheURL{$k};
	    }
	}
    }
  $self->{logger}->debug("End of sort of URL");
  $self->{savespace}{'nsCacheURLTableEntries'} = $#nsCacheURL + 1;
  for (my $i = 0 ; $i <= $#nsCacheURL; $i ++) {
    my $i1 = $i + 1;
    $self->{savespace}{'nsCacheURLIndex.'.$i1} = $i1;
    foreach my $k (keys %{$nsCacheURL[$i]}) {
      if ($k ne 'ResponseTime') {
	$self->{savespace}{'nsCacheURL' . $k . '.'. $i1} = ${$nsCacheURL[$i]}{$k};
      } else {
	if (${$nsCacheURL[$i]}{$k}->count == 0) {
	  $self->{savespace}{'nsCacheURLResponseTimeEntries'. '.'. $i1} = 0;
	  $self->{savespace}{'nsCacheURLAvgResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheURLStdDeviationResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheURLMinResponseTime'. '.'. $i1} =  0;
	  $self->{savespace}{'nsCacheURLMaxResponseTime'. '.'. $i1} =  0;
	} else {
	  $self->{savespace}{'nsCacheURLResponseTimeEntries'. '.'. $i1} = ${$nsCacheURL[$i]}{'ResponseTime'}->count();
	  $self->{savespace}{'nsCacheURLAvgResponseTime'. '.'. $i1} = ${$nsCacheURL[$i]}{'ResponseTime'}->mean();
	  $self->{savespace}{'nsCacheURLStdDeviationResponseTime'. '.'. $i1} = defined(${$nsCacheURL[$i]}{'ResponseTime'}->standard_deviation()) ? ${$nsCacheURL[$i]}{'ResponseTime'}->standard_deviation(): 0;
	  $self->{savespace}{'nsCacheURLMinResponseTime'. '.'. $i1} = ${$nsCacheURL[$i]}{'ResponseTime'}->min();
	  $self->{savespace}{'nsCacheURLMaxResponseTime'. '.'. $i1} = ${$nsCacheURL[$i]}{'ResponseTime'}->max();
	}
      }
    }
  }

  $self->properties($self->savespace);
  delete $self->{properties}{responseTime};
  delete $self->{properties}{dnsResponseTime};
}

1;

=head1 NAME

LogparserDriver

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 REQUIREMENTS AND LIMITATIONS

=head1 OPTIONS

=head1 BUGS

=head1 TODO

=over 8

=item * document logger.

=back

=head1 SEE ALSO

=head1 AUTHOR

Nito at Qindel dot ES -- 7/9/2006

=head1 COPYRIGHT & LICENSE

Copyright 2007 by Qindel Formacion y Servicios SL, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/SNMP/LogParserDriver.pm
#
# $Id: LogParserDriver.pm,v 1.3 2006/09/25 06:28:44 nito Exp $
#
# Simple class that counts he number of lines of a file
#
# Nito@Qindel.ES -- 7/9/2006
package SNMP::LogParserDriver;

use warnings;
use strict;
use Log::Log4perl qw(:easy);;
use Data::Dumper;

=head1 NAME

SNMP:LogParserDriver

=head1 SYNOPSIS

This is base class for the usage of the logparser
parsing. Usually you will implement a subclass of
this class implementing the following methods:

=over 8

=item * new


You can define whatever new parameters you want
in this method, but usually only the object variable
B<pattern> (the regular expression to match in the file)
is needed

=item * evalBegin


This is an extract of code which initializes all the variables
that you might need during the run of logparser

=item * evalIterate


This method will be invoked for each line of the log file.

=item * evalEnd


This method usually records in the B<properties> hash reference
the key, value pairs that you want ot record in the output.

=back

=head1 DESCRIPTION

For a more detailed description on how to use this class
please check B<logparser>.

This document is a detailed reference of the different
methods provided in this class

=head1 METHODS

=head2 new

Receives (at the moment) no arguments. It creates an object
of SNMP::LogParserDriver

=cut

# Class constructor
sub new {
  my $class = shift;
  my $self = {};
  $self = {
	   name => undef,
	   logfile => undef,
	   pattern => undef,
	   savespace => undef,
	   workspace => undef,
	   properties => undef,
	   logger => undef
	  };
  bless ($self, $class);
  $self->name($class);
  $self->pattern('^.*$');
  if(!(Log::Log4perl->initialized())) {
    Log::Log4perl->easy_init($ERROR);
  }
  $self->logger(Log::Log4perl->get_logger("logparser.$class"));
  $self->{logger}->info("Created object of class $class");
  return $self;
}

=head2 name

Permits to retrieve or set the name of the object as in

 print "name: ".$self->name;

or

 $self->name("myname");

=cut

# The name of this object
sub name {
  my $self = shift;
  $self->{name} = shift if (@_);
  return $self->{name};
}

=head2 logfile

Permits to retrieve or set the logfile of the object as in

 print "logfile: ".$self->logfile;

or

 $self->logfile("/var/log/logfile");

=cut

# The log file that should be parsed
sub logfile {
  my $self = shift;
  $self->{logfile} = shift if (@_);
  return $self->{logfile};
}

=head2 pattern

Permits to retrieve or set the regular expression pattern of the object as in

 print "pattern: ".$self->pattern;

or

 $self->pattern("^myregex$");

=cut

# The regular expression that should be matched
sub pattern {
  my $self = shift;
  $self->{pattern} = shift if (@_);
  return $self->{pattern};
}

=head2 savespace

Permits to retrieve or set the savespace of the object as in

 print "savespace: ".Dumper($self->savespace);

or

 $self->savespace(\%mysavespace);

The savespace is usually a reference to a hash file. The
value of the savespace variable will be restored in the beginning
of each invocation of B<logparser> and saved at the end of the
invocation. Tipically things like counters are saved in each
invocation of the savespace.

=cut

# The savespace that should be saved across sessions
sub savespace {
  my $self = shift;
  $self->{savespace} = shift if (@_);
  return $self->{savespace};
}

=head2 workspace

Permits to retrieve or set a variable that will be maintained across iterative parsings
of the log, but will not be maintained across logparser invocations

 print "workspace: ".Dumper($self->workspace);

or

 $self->workspace(\%workspace);

The workpace is usually a reference to a hash file. The
value of the workspace variable will only last for each invocation
of B<logparser>. Tipically things like local variable are saved in
in the workspace.

=cut

# A variable that will be maintained across iterative parsings
# of the log, but will not be maintained across logparser invocations
sub workspace {
  my $self = shift;
  $self->{workspace} = shift if (@_);
  return $self->{workspace};
}

=head2 properties

Permits to retrieve or set the properties of the object as in

 print "properties: ".$self->properties;

or

 $self->properties(\%myproperties);

The properties of an object are the key/value pairs that will
be stored in the B<propertiesFile> of the B<logparser.conf>
file.

=cut

# The properties that should be set in the output file
sub properties {
  my $self = shift;
  $self->{properties} = shift if (@_);
  return $self->{properties};
}

=head2 logger

Permits to retrieve or set the logger of the object as in

 print $self->logger->debug("debug message");

or

 $logger = Log::Log4perl->get_logger(LOG_TAG);
 $self->logger($logger);

=cut

# Sets the logger object to allow for logging
sub logger {
  my $self = shift;
  $self->{logger} = shift if (@_);
  return $self->{logger};
}


=head2 evalBegin

This method is usually overriden in the child class.
This method will be invoked before any line of the
file is parsed.

Typically counters will be initialised here if they
have not been initialised before:

  $self->{savespace}{counter} = 0 if (!exists($self->{savespace}{counter}));

=cut

# This will be invoked before the first parsing of the log
sub evalBegin {
  my $self = shift;
  $self->{savespace}{counter} = 0 if (!exists($self->{savespace}{counter}));
}

=head2 evalIterate

This method is usually overriden in the child class.
This method will be invoked for each line of the 
log that should be parsed.

It receives as an argument the line of the log to be parsed.

One typical implementation could be:

 sub evalIterate {
  my $self = shift;
  my ($line) = @_;
  my $pattern = $self->{pattern};
  if ($line =~ /$pattern/) {
    $self->{savespace}{counter} ++;
  }
 }


=cut

# This will be invoked whenever the pattern matches
# the log line parsed
# Input:
# - The line to be parsed
# Output:
# - 1 if the line has matched the regular expression and 0 otherwise
sub evalIterate {
  my $self = shift;
  my ($line) = @_;
  my $pattern = $self->{pattern};
  if ($line =~ /$pattern/) {
    $self->{savespace}{counter} ++;
  }
}

=head2 evalEnd

This method is usually overriden in the child class.
This method will be invoked after all the lines
have been parsed.

Typically the properties elements will be initialised here

=cut

# This will be invoked after the last log line has been parsed
sub evalEnd {
  my $self = shift;
  $self->properties($self->savespace());
}

1;


=head1 REQUIREMENTS AND LIMITATIONS

=head1 BUGS

=head1 TODO

=over 8

=item * document logger.

=back

=head1 SEE ALSO

=head1 AUTHOR

Nito@Qindel.es -- 7/9/2006


Build.PL

use Module::Build;
my $build = Module::Build->new(
    module_name        => 'SNMP::LogParser',
    sign               => 1,
    license            => 'perl',
    dist_author        => 'Nito Martinez <nito@qindel.es>',
    create_makefile_pl => 'passthrough',
    dist_version_from  => 'lib/SNMP/LogParser.pm',
    script_files       => ['bin/logparser'],
    install_path       => {
        'etc'  => '/etc/logparser',
        'work' => '/var/lib/logparser',
        'log'  => '/var/log/logparser'
    },
    etc_files => {
        'etc/log4perl.conf'  => 'etc/log4perl.conf',
        'etc/logparser.conf' => 'etc/logparser.conf'
    },
    log_files => { 'log/logparser.log' => 'log/logparser.log' },
    work_files =>
      { 'work/logparser.properties' => 'work/logparser.properties' },
    requires => {
        'Test::More'                 => 0,
        'Test::SharedFork'           => 0,
        'Log::Log4perl'              => 0,
        'Log::Dispatch::FileRotate'  => 0,
        'Proc::ProcessTable'         => 0,
        'File::Spec'                 => 0,
        'File::stat'                 => 0,
        'File::Temp'                 => 0,
        'File::Path'                 => 1.05,
        'Pod::Usage'                 => 0,
        'Test::Pod::Coverage'        => 1.04,
        'Test::Pod'                  => 1.14,
        'Proc::ProcessTable'         => 0,
        'Statistics::Descriptive'    => 0,
        'Storable'                   => 0,
        'Config::Properties::Simple' => 0.14,
        'POSIX'                      => 0,
        'File::Temp'                 => 0,
        'parent'                     => 0,
        'lib'                        => 0,
        'IPC::Run3'                  => 0,
    },
);

$build->create_makefile_pl('traditional');
$build->add_build_element('etc');
$build->add_build_element('log');
$build->add_build_element('work');
$build->create_build_script;
